src/index.js

```diff
@@ -3,103 +3,105 @@ const http = require('http')
 
-app.get('/rooms', async (req, res) => {
+app.get('/rooms', (req, res) => {
     try {
-        const rooms = getRooms()
-        res.send({ rooms })
+        const rooms = roomRepo.all
+        res.send({ rooms: rooms.map((r) => r.title) })
     } catch (error) {
+        console.error(error)
         res.status(500).send({ error: error.message })
     }
 })
 
 io.on('connection', (socket) => {
     console.log('New WebSocket connection.')
+    let context = {
+        user: null,
+        room: null,
+    }
 
-    socket.on('join', ({ username, room }, callback) => {
-        const { error, user } = addUser({ id: socket.id, username, room })
+    let generateMessage
+    let generateLocationMessage
 
-        if (error) {
-            return callback(error)
-        }
+    socket.on('join', async ({ username, room: roomTitle }, callback) => {
+        const { error, user, room } = await userToRoomJoiner.call({
+            username,
+            roomTitle,
+        })
+        if (error) return callback(error)
 
-        socket.join(user.room)
+        context = {
+            user,
+            room,
+        }
+        generateMessage = makeMessageGenerator(user)
+        generateLocationMessage = makeMessageGenerator(
+            user,
+            (location) => `${location}`,
+            'url'
+        )
 
-        socket.emit('message', generateMessage('Admin', 'Welcome!'))
-        socket.broadcast
-            .to(user.room)
-            .emit(
-                'message',
-                generateMessage('Admin', `${user.username} has joined!`)
-            )
+        socket.join(room)
+        socket.emit('message', generateWelcome())
+        socket.broadcast.to(room).emit('message', generateJoined(user))
 
-        io.to(user.room).emit('roomData', {
-            room: user.room,
-            users: getUsersInRoom(user.room),
-        })
+        const roomWithUsers = await roomDataFetcher.call(room)
+        io.to(room).emit('roomData', roomWithUsers)
 
         callback()
     })
 
     socket.on('sendMessage', (message, callback) => {
-        const user = getUser(socket.id)
-        const filter = new Filter()
         if (filter.isProfane(message)) {
             return callback('Profanity is not allowed.')
         }
 
-        io.to(user.room).emit(
-            'message',
-            generateMessage(user.username, message)
-        )
+        io.to(context.room).emit('message', generateMessage(message))
         callback()
     })
 
-    socket.on('getLocation', (location, callback) => {
-        const user = getUser(socket.id)
-        io.to(user.room).emit(
+    socket.on('getLocation', (coords, callback) => {
+        io.to(context.room).emit(
             'sendLocation',
-            generateLocationMessage(
-                user.username,
-                `https://google.com/maps?q=${location.latitude},${location.longitude}`
-            )
+            generateLocationMessage(new Location(coords))
```

Comment:
How are you passing stuff into `generateLocationMessage`?
I mean I though that it holds a value of what `makeMessageGenerator` returns
`
        generateLocationMessage = makeMessageGenerator(
            user,
            (location) => `${location}`,
            'url'
        )
`

Reply:
`generateLocationMessage` it is a function returned from `makeMessageGenerator`

Reply:
Do I understand correctly that `new Location(coords)` is the `payload` that from the `makeMessageGenerator` method, right?

And because of the closure, this new `generateLocationMessage` has knows what to put into `author.username` and `[payloadKey]`?

`
({
        username: author.username,
        [payloadKey]: process(payload),
        createdAt: Date.now(),
    })
`

Reply:
yes